File: packages/api/myvoice/api/packs.py
"""REST endpoints for browsing style packs."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Request

router = APIRouter(prefix="/api/packs", tags=["packs"])


@router.get("")
def list_packs(request: Request) -> list[dict[str, Any]]:
    """Return summaries of all discovered packs."""
    store = request.app.state.pack_store
    return [
        {
            "slug": info.slug,
            "name": info.name,
            "version": info.version,
            "valid": info.valid,
            "error_count": len(info.errors),
        }
        for info in (store.get(slug) for slug in store.slugs())
        if info is not None
    ]


@router.get("/{slug}")
def get_pack(slug: str, request: Request) -> dict[str, Any]:
    """Pack detail: manifest summary + content counts + validation status."""
    store = request.app.state.pack_store
    info = store.get(slug)
    if info is None:
        raise HTTPException(status_code=404, detail=f"pack '{slug}' not found")

    detail: dict[str, Any] = {
        "slug": info.slug,
        "name": info.name,
        "version": info.version,
        "root_path": str(info.root_path),
        "valid": info.valid,
        "errors": [{"path": e.path, "message": e.message} for e in info.errors],
    }
    # If valid, include manifest counts
    from myvoice.validate import validate_pack

    result = validate_pack(info.root_path)
    if result.manifest is not None:
        m_obj = result.manifest
        detail["author"] = m_obj.pack.author
        detail["description"] = m_obj.pack.description
        detail["persona"] = {
            "identity": m_obj.persona.identity,
            "one_line": m_obj.persona.one_line,
        }
        detail["counts"] = {
            "banished_words": len(m_obj.banished.words),
            "banished_phrases": len(m_obj.banished.phrases),
            "permitted_exceptions": len(m_obj.banished.permitted_exceptions),
            "formats": len(m_obj.formats),
            "samples": len(m_obj.samples),
            "bios": len(m_obj.bios),
        }
    return detail


@router.get("/{slug}/manifest")
def get_manifest(slug: str, request: Request) -> dict[str, Any]:
    """Full manifest as JSON."""
    import yaml

    store = request.app.state.pack_store
    info = store.get(slug)
    if info is None:
        raise HTTPException(status_code=404, detail=f"pack '{slug}' not found")

    manifest_path = info.root_path / "stylepack.yaml"
    raw = yaml.safe_load(manifest_path.read_text(encoding="utf-8"))
    assert isinstance(raw, dict)
    return raw

File: packages/web/src/api/packs.ts
import { apiFetch } from "./client";

export interface PackSummary {
  slug: string;
  name: string;
  version: string;
  valid: boolean;
  error_count: number;
}

export interface PackDetail extends PackSummary {
  root_path: string;
  errors: { path: string; message: string }[];
  author?: string;
  description?: string | null;
  persona?: { identity: string; one_line: string };
  counts?: {
    banished_words: number;
    banished_phrases: number;
    permitted_exceptions: number;
    formats: number;
    samples: number;
    bios: number;
  };
}

export async function listPacks(init?: RequestInit): Promise<PackSummary[]> {
  return apiFetch<PackSummary[]>("/api/packs", init);
}

export async function getPack(slug: string, init?: RequestInit): Promise<PackDetail> {
  return apiFetch<PackDetail>(`/api/packs/${encodeURIComponent(slug)}`, init);
}

export async function getManifest(
  slug: string,
  init?: RequestInit,
): Promise<Record<string, unknown>> {
  return apiFetch<Record<string, unknown>>(`/api/packs/${encodeURIComponent(slug)}/manifest`, init);
}

export async function getPackFile(slug: string, path: string, init?: RequestInit): Promise<string> {
  // Bypass apiFetch's JSON expectation — file endpoint returns text/plain.
  const response = await fetch(`/api/packs/${encodeURIComponent(slug)}/files/${path}`, init);
  if (!response.ok) {
    throw new Error(`HTTP ${response.status} on file ${path}`);
  }
  return response.text();
}

export interface CreatePackRequest {
  slug: string;
  name: string;
  author: string;
  persona_identity: string;
  persona_one_line: string;
  persona_tone?: string;
  version?: string;
  description?: string;
}

export async function createPack(req: CreatePackRequest): Promise<PackSummary> {
  return apiFetch<PackSummary>("/api/packs", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export async function deletePack(slug: string): Promise<void> {
  const res = await fetch(`/api/packs/${encodeURIComponent(slug)}`, { method: "DELETE" });
  if (!res.ok && res.status !== 204) {
    throw new Error(`HTTP ${res.status} deleting pack ${slug}`);
  }
}
